### ISSUE TITLE
feat(gp): transformed Gaussian processes via a warped likelihood

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/pyrox#203]
## Summary

Add support for **transformed Gaussian processes** (Maroñas et al., AISTATS 2021, [arXiv:2011.01596](https://arxiv.org/abs/2011.01596)): a GP composed with an elementwise monotone warp `G`, giving non-Gaussian marginals from a standard GP prior.

The entire feature is one `Likelihood` subclass plus a quadrature helper. **No change to `svgp_elbo`, `svgp_factor`, any `Guide`, `SparseGPPrior`, the inducing-feature machinery, or `gaussx`.**

## Why it is this small

A transformed GP is an SVGP whose likelihood composes a warp:

$$
\mathcal{L} = \sum_n \mathbb{E}_{q(f_n)}\big[\log p(y_n \mid G(f_n))\big] - \mathrm{KL}[q(u) \Vert p(u)]
$$

That is character-for-character the ELBO `svgp_elbo` already computes (`_inference.py:88-91`), with `log p(y|f)` replaced by `log p(y|G(f))`. **The warp does not appear in the KL term at all** — the variational family is over inducing values of the *base* GP, so `KL[q(u)‖p(u)]` is untouched.

Confirmed against the reference implementation: its plain-SVGP class is literally its transformed-GP class instantiated with an identity flow.

Consequence: every existing inference path in `pyrox_gp` — SVGP, Laplace, EP, posterior linearization — inherits the capability for free, because they all consume `Likelihood.log_prob`.

## Dependency: `flowjax`

The warp types against `flowjax.bijections.AbstractBijection`. Without it this issue would have to define a bespoke bijection base class, which is strictly worse — flowjax bijections are monotone by construction and several initialize at the exact identity.

```toml
# packages/pyrox-gp/pyproject.toml
dependencies = [
    # ... existing ...
    "flowjax>=0.9",   # AbstractBijection protocol, Invert / Chain /
                      # RationalQuadraticSpline. Shares lineax, equinox, jax
                      # and jaxtyping with pyrox-gp already; net new
                      # transitive deps are just paramax and tqdm.
]
```

Optionally also the `[flows]` extra for `gauss_flows.MixtureGaussianCDF` (the recommended warp — see below). **That extra may already exist** if the latent-factor issues have landed; add it only if absent, do not duplicate:

```toml
[project.optional-dependencies]
flows = ["gauss-flows @ git+https://github.com/jejjohnson/gauss_flows.git@<pin>"]
```

## ⚠️ Warp choice is a numerics decision

The expected log-likelihood is evaluated by Gauss-Hermite quadrature, which converges **spectrally only for analytic integrands**. A rational-quadratic spline is `C²` at its knots but not `C^∞`, so the guarantee degrades — and because knot locations are fixed while quadrature nodes move with the order, the error *oscillates* rather than decreasing.

*Measured* (`y=1.4, μ=0.3, var=0.7, σ²=0.1`, warps perturbed off identity):

| GH order | sinh-arcsinh (analytic) | Δ | RQ spline (piecewise) | Δ |
|---|---|---|---|---|
| 8 | −22.36677365 | — | −8.86221646 | — |
| 16 | −22.37280485 | 6.0e−03 | −9.12091560 | 2.6e−01 |
| 32 | −22.37290747 | 1.0e−04 | −9.11949592 | 1.4e−03 |
| 64 | −22.37290635 | 1.1e−06 | −9.09250739 | **2.7e−02** |
| 128 | −22.37290635 | 3.3e−09 | −9.09399499 | 1.5e−03 |
| 256 | −22.37290635 | **8.6e−13** | −9.09744243 | 3.5e−03 |

And the smooth `gauss_flows` alternative, `MixtureGaussianCDF` (a mixture of Gaussian CDFs followed by a probit — analytic everywhere):

| GH order | 32 | 64 | 128 | 256 |
|---|---|---|---|---|
| Δ | 2.5e−04 | 2.6e−06 | 1.3e−08 | **7.1e−12** |

**Document `MixtureGaussianCDF` as the recommended warp.** Splines still work — `RationalQuadraticSpline` is the exact identity at init, which is a real convenience — but their `~3e-3` error floor must be stated, and users must not be told to "raise the order until it converges", because for splines that diagnostic is invalid and raising the order can make it worse.

Two further practical notes:

- The warped integrand has **heavy tails**: `log N(y|G(f),σ²) ≈ −G(f)²/2σ²` with `G` sinh-like. Measured integrand sd `32.8`, min `−1054`. Monte Carlo needs ~4M samples to match Gauss-Hermite at order 20. **Gauss-Hermite must be the documented default**; `MonteCarloIntegrator` should carry a warning.
- `numpy.polynomial.hermite_e.hermegauss` overflows to `NaN` at order 512. Validate or cap the order.

> **If you have also read the latent-factor warped-regression issue:** its recommendation is the *opposite* (splines preferred). That is deliberate. Its collapsed likelihood is exact — no quadrature in the training loop — so smoothness is irrelevant there and closed-form invertibility dominates. Do not harmonise the two.

## Files to create / change

| File | Change |
|---|---|
| `packages/pyrox-gp/src/pyrox_gp/_warped.py` | **new** — the code below |
| `packages/pyrox-gp/pyproject.toml` | `flowjax>=0.9`; `[flows]` extra if absent |
| `packages/pyrox-gp/src/pyrox_gp/__init__.py` | export `WarpedGaussianLikelihood`, `warped_predictive_moments` |
| `packages/pyrox-gp/tests/gp/test_warped.py` | **new** |
| `docs/api/gp.md` | new entries |

## Reference implementation

Validated — see "Validation evidence". Google-style docstrings with `$...$` math; **no Sphinx/RST markup**.

```python
# packages/pyrox-gp/src/pyrox_gp/_warped.py
"""Transformed Gaussian processes: a GP composed with a monotone warp.

A transformed GP (Maronas et al., AISTATS 2021) is an SVGP whose likelihood
composes an elementwise monotone map $G$:

$$
\\mathcal{L} = \\sum_n \\mathbb{E}_{q(f_n)}[\\log p(y_n \\mid G(f_n))]
             - \\mathrm{KL}[q(u) \\Vert p(u)]
$$

The warp appears only in the expected log-likelihood -- the KL term is
untouched -- so this composes with every existing inference path in
`pyrox_gp` without modification.
"""

from __future__ import annotations

import jax
import jax.numpy as jnp
import numpy as np
import numpyro.distributions as nd
from flowjax.bijections import AbstractBijection
from jaxtyping import Array, Float

from pyrox_gp._protocols import Likelihood


def _apply_warp(warp: AbstractBijection, f: Float[Array, " ..."]) -> Array:
    """Apply a scalar bijection elementwise, preserving shape.

    Shape-agnostic on purpose: `pyrox_gp.svgp_elbo` integrates
    per point and calls ``log_prob`` with ``f`` of shape ``(1,)``, while the
    predictive path and the advanced inference strategies pass full ``(N,)``
    arrays.
    """
    if warp.shape == ():
        fn = warp.transform
    elif warp.shape == (1,):
        fn = lambda v: warp.transform(v[None])[0]
    else:
        raise ValueError(
            f"Warp event shape must be () or (1,); got {warp.shape}. "
            "A transformed GP warps each scalar function value independently."
        )
    return jnp.reshape(jax.vmap(fn)(jnp.reshape(f, (-1,))), jnp.shape(f))


class WarpedGaussianLikelihood(Likelihood):
    r"""$p(y \mid f) = N(y \mid G(f), \sigma^2)$ -- the transformed-GP model.

    The warp is a **child module**, so its parameters are ordinary trainable
    leaves. Passing a warp through a lambda to
    [`DistLikelihood`][pyrox_gp.DistLikelihood] does *not* work: its
    ``dist_fn`` is a static field, so the warp is silently frozen and never
    trains.

    The expected log-likelihood has no closed form, so
    [`svgp_elbo`][pyrox_gp.svgp_elbo] requires an integrator.

    !!! warning "Prefer Gauss-Hermite, and prefer a smooth warp"
        The warped integrand has heavy tails, so Monte Carlo integration is
        high variance -- roughly 4M samples to match Gauss-Hermite at order
        20. Gauss-Hermite in turn converges spectrally only for *analytic*
        warps: piecewise ones such as ``RationalQuadraticSpline`` stall
        around a ``3e-3`` error floor and their error is **non-monotone** in
        quadrature order, so raising the order is not a valid convergence
        check. ``gauss_flows.MixtureGaussianCDF`` is smooth and reaches
        machine precision.

    Attributes:
        warp: Any ``flowjax.bijections.AbstractBijection`` with event shape
            ``()`` or ``(1,)``.
        noise_var: Observation noise variance $\sigma^2$.

    Examples:
        >>> import jax.numpy as jnp
        >>> from flowjax.bijections import RationalQuadraticSpline
        >>> lik = WarpedGaussianLikelihood(
        ...     warp=RationalQuadraticSpline(knots=8, interval=4.0),
        ...     noise_var=jnp.asarray(0.1),
        ... )
        >>> float(lik.log_prob(jnp.zeros(3), jnp.zeros(3))) < 0.0
        True
    """

    warp: AbstractBijection
    noise_var: Float[Array, ""]

    def log_prob(
        self,
        f: Float[Array, " ..."],
        y: Float[Array, " ..."],
    ) -> Float[Array, ""]:
        r"""Sum of per-point Gaussian log-densities about $G(f)$."""
        g = _apply_warp(self.warp, f)
        return nd.Normal(g, jnp.sqrt(self.noise_var)).log_prob(y).sum()


def warped_predictive_moments(
    lik: WarpedGaussianLikelihood,
    f_loc: Float[Array, " N"],
    f_var: Float[Array, " N"],
    *,
    order: int = 32,
) -> tuple[Float[Array, " N"], Float[Array, " N"]]:
    r"""Moments of $y = G(f) + \epsilon$ under $q(f)$, by quadrature.

    $$
    m_1 = \mathbb{E}_{q(f)}[G(f)], \qquad
    m_2 = \sigma^2 + \mathbb{E}_{q(f)}[G(f)^2] - m_1^2
    $$

    Note $m_1 \neq G(\mathbb{E}[f])$ whenever $G$ is nonlinear -- warping the
    posterior mean is the natural-looking mistake and is badly biased for a
    skewed warp.

    Args:
        lik: The warped likelihood.
        f_loc: Posterior means of the base GP, shape ``(N,)``.
        f_var: Posterior marginal variances of the base GP, shape ``(N,)``.
        order: Gauss-Hermite nodes. Values above ~256 are numerically
            unreliable.

    Returns:
        Tuple of ``(mean, variance)`` in observation space, both ``(N,)``.
    """
    if not 1 <= order <= 256:
        raise ValueError(f"order must be in [1, 256]; got {order}.")
    x, w = np.polynomial.hermite_e.hermegauss(order)
    x = jnp.asarray(x)
    w = jnp.asarray(w) / np.sqrt(2.0 * np.pi)
    fs = f_loc[None, :] + jnp.sqrt(f_var)[None, :] * x[:, None]
    g = _apply_warp(lik.warp, fs)
    m1 = jnp.sum(w[:, None] * g, axis=0)
    m2 = lik.noise_var + jnp.sum(w[:, None] * g**2, axis=0) - m1**2
    return m1, m2
```

## Usage

```python
from flowjax.bijections import RationalQuadraticSpline
from gaussx import GaussHermiteIntegrator
from pyrox_gp import (
    RBF, SparseGPPrior, WhitenedGuide, WarpedGaussianLikelihood, svgp_elbo,
)

lik = WarpedGaussianLikelihood(
    warp=RationalQuadraticSpline(knots=8, interval=4.0),   # exact identity at init
    noise_var=jnp.asarray(0.1),
)
elbo = svgp_elbo(prior, guide, lik, X, y,
                 integrator=GaussHermiteIntegrator(order=32))
```

Recommended warp, with the `flows` extra. `MixtureGaussianCDF` starts as a pure scale of `1 / (softplus(0) + 5e-3) = 1.432363`, so compose the exact inverse scale for identity init:

```python
import jax
from flowjax.bijections import Affine, Chain
from gauss_flows import MixtureGaussianCDF

IDENTITY_SCALE = float(jax.nn.softplus(0.0) + 5e-3)
warp = Chain([
    MixtureGaussianCDF(n_components=8, shape=(1,)),
    Affine(loc=jnp.zeros(1), scale=jnp.full((1,), IDENTITY_SCALE)),
])
```

*Verified:* `max|G(f) − f| = 1.2e-7`. **`Chain` requires matching event shapes** — the `Affine` must be `(1,)`, not scalar, or construction raises.

## Acceptance criteria

New file `packages/pyrox-gp/tests/gp/test_warped.py`, with `jax.config.update("jax_enable_x64", True)` at module top (precedent: `tests/gp/test_markov_gp.py:27`).

- [ ] **Identity warp reduces exactly to the Gaussian ELL.** With
      `RationalQuadraticSpline(knots=8, interval=4.0)` (exact identity at init),
      the quadrature ELL equals the closed form
      `−0.5·log(2πσ²) − ((y−μ)² + var)/(2σ²)` summed over points, to `1e-12`.
- [ ] **Non-identity warp vs Monte Carlo.** Perturb the warp parameters; compare
      Gauss-Hermite against ≥1e5 Monte Carlo samples and assert agreement within
      2 MC standard errors. Mark `@pytest.mark.slow`.
- [ ] **Quadrature convergence is warp-dependent.** Assert that an analytic warp
      (`gauss_flows.MixtureGaussianCDF`) converges — successive-order deltas
      strictly decreasing past order 32 and below `1e-9` by order 128 — while a
      spline warp does **not** reach that floor. This is what stops the §"warp
      choice" finding being silently lost; without it someone will later
      "simplify" the docs. Skip the smooth half when `gauss_flows` is absent.
- [ ] **Gradients reach the warp.** `eqx.filter_grad` produces a **non-empty**,
      finite gradient over the warp's leaves and over `noise_var`. This is the
      direct contrast with the `DistLikelihood` trap in [T1].
- [ ] **Shape-agnostic application.** `log_prob` is correct at `f` shapes `(1,)`,
      `(N,)`, and `(S, N)`, and per-point application matches batched
      application exactly. `(1,)` is what `svgp_elbo` actually passes.
- [ ] **End to end through `svgp_elbo`.** One SVI step on a small problem with
      `GaussHermiteIntegrator`; assert finite loss and that the warp parameters
      moved.
- [ ] **Predictive moments vs Monte Carlo**, 2% relative. Mark `slow`.
- [ ] **`E[G(f)] ≠ G(E[f])`** for a non-identity warp — asserts the bias is real,
      so nobody replaces the quadratic with a point evaluation later.
- [ ] **Invalid event shape** raises `ValueError` naming the expected shape.
- [ ] **`order` out of range** raises `ValueError` (`hermegauss` returns `NaN` at
      512).
- [ ] **`eqx.filter_jit` works.** Note that **plain `jax.jit` fails** on flowjax
      bijections — spline knot positions carry string leaves
      (`include_endpoints`), producing a confusing
      "problematic value is of type `str`" error. Worth an explicit test so the
      requirement is discoverable.

## Validation evidence

Executed against the reference implementation:

| Check | Result |
|---|---|
| Identity-init warp ⇒ closed-form Gaussian ELL | `1.4e-14` |
| GH vs 300k-sample MC (spline warp) | within `1.1` MC stderr |
| Gradients into warp (3 leaves) + `noise_var` | all finite, non-empty |
| Monotonicity under randomised parameters (RQS, `MixtureGaussianCDF`) | strictly increasing, no constraint code needed |
| Predictive moments vs MC | mean err `0.0044`, var err `0.0083` |
| `G(E[f])` vs `E[G(f)]` gap | `0.079` |
| Shape-agnostic apply at `(1,)`, `(7,)`, `(3,4)` | all correct; per-point ≡ batched to `0.0` |
| `jax.jit` | **fails**; `eqx.filter_jit` works |

## Out of scope

- **Input-dependent (non-stationary) warps** → issue [T3]. Requires an optional
  `X` on the `Likelihood` protocol; `_apply_warp` above deliberately rejects
  conditional bijections implicitly by never passing a condition.
- **Classic warped GP** (Snelson 2003 — warp the *observations*, needing
  `log|dT/dy|`). A different model; smaller, separate feature.
- **Multi-output warps.** Composing with `LMCKernel` / `OILMMKernel` is open
  territory; the reference implementation only supports one independent warp per
  output and asserts the general case away.
- **Data-driven warp initialization** (`gauss_flows.fit_rbig`,
  `MixtureGaussianCDF.from_data`). Useful later; identity init is enough to land
  this.

## Note on invertibility

The paper motivates `G` as a normalizing flow, implying bijectivity, but the ELBO
$\log p(y) \ge \mathbb{E}_{q(f)}[\log p(y \mid G(f))] - \mathrm{KL}$
is valid for **any** `G` — the model is just "a GP through a nonlinear link", and
the training path never inverts the warp. Monotonicity buys interpretation and
identifiability (it stops `G` folding the latent space and making `f`
unidentifiable), not correctness. So this issue needs neither `inverse` nor
`log_abs_det_jacobian`, which is why the API surface is as small as it is.

## Before reporting done

```bash
uv run pytest -q
uv run --group lint ruff check .          # note: '.', not 'packages/'
uv run --group lint ruff format --check .
uv run --group typecheck ty check packages/pyrox-gp/src/pyrox_gp
```
